index/segment/fst/writer.go

```diff
@@ -0,0 +1,361 @@
+
+	intEncoder      *encoding.Encoder
+	postingsEncoder *pilosa.Encoder
+	fstWriter       *fstWriter
+
+	metadata            []byte
+	docsDataFileWritten bool
+	postingsFileWritten bool
+	fstTermsFileWritten bool
+	postingsOffsets     *postingsOffsetsMap
+	fstTermsOffsets     *fstTermsOffsetsMap
+}
+
+// NewWriter returns a new writer.
+func NewWriter() Writer {
+	return &writer{
+		intEncoder:      encoding.NewEncoder(defaultInitialIntEncoderSize),
+		postingsEncoder: pilosa.NewEncoder(),
+		fstWriter:       newFSTWriter(),
+		postingsOffsets: newPostingsOffsetsMap(defaultInitialPostingsOffsetsMapSize),
+		fstTermsOffsets: newFSTTermsOffsetsMap(defaultInitialFSTTermsOffsetsMapSize),
+	}
+}
+
+func (w *writer) Clear() {
+	w.seg = nil
+
+	w.intEncoder.Reset()
+	w.postingsEncoder.Reset()
+	w.fstWriter.Clear()
+
+	w.metadata = nil
+	w.docsDataFileWritten = false
+	w.postingsFileWritten = false
+	w.fstTermsFileWritten = false
+	w.postingsOffsets.Reset()
+	w.fstTermsOffsets.Reset()
+}
+
+func (w *writer) Reset(s segment.MutableSegment) error {
+	w.Clear()
+
+	numDocs := s.Size()
+	metadata := defaultV1Metadata()
+	metadata.NumDocs = numDocs
+	metadataBytes, err := metadata.Marshal()
+	if err != nil {
+		return err
+	}
+
+	w.metadata = metadataBytes
+	w.seg = s
+	return nil
+}
+
+func (w *writer) MajorVersion() int {
+	return majorVersion
+}
+
+func (w *writer) MinorVersion() int {
+	return minorVersion
+}
+
+func (w *writer) Metadata() []byte {
+	return w.metadata
+}
+
+func (w *writer) WriteDocumentsData(iow io.Writer) error {
+	return nil
+}
+
+func (w *writer) WriteDocumentsIndex(iow io.Writer) error {
+	if !w.docsDataFileWritten {
+		return fmt.Errorf("documents data file has to be written before documents index file")
+	}
+	return nil
+}
+
+func (w *writer) WritePostingsOffsets(iow io.Writer) error {
+	currentOffset := uint64(0)
+
+	// retrieve known fields
+	fields, err := w.seg.Fields()
+	if err != nil {
+		return err
+	}
+
+	// for each known field
+	for _, f := range fields {
+		// retrieve known terms for current field
+		terms, err := w.seg.Terms(f)
+		if err != nil {
+			return err
+		}
+
+		// for each term corresponding to the current field
+		for _, t := range terms {
+			// retrieve the postings list for this (field, term) combination
+			pl, err := w.seg.MatchTerm(f, t)
+			if err != nil {
+				return err
+			}
+
+			// serialize the postings list
+			w.postingsEncoder.Reset()
+			postingsBytes, err := w.postingsEncoder.Encode(pl)
+			if err != nil {
+				return err
+			}
+
+			n, err := w.writePayloadSizeMagicNumber(iow, postingsBytes)
+			if err != nil {
+				return err
+			}
+
+			// update offset with the number of bytes we've written
+			currentOffset += n
+
+			// track current offset as the offset for the current field/term
+			w.addPostingsOffset(currentOffset, f, t)
+		}
+	}
+
+	w.postingsFileWritten = true
+	return nil
+}
+
+func (w *writer) WriteFSTTerms(iow io.Writer) error {
+	if !w.postingsFileWritten {
```

Comment:
Should also test the other files written too?

Reply:
We can rely on the check being correct transitively yea? As long as each file checks it's requirement, we get the same total guarantee.